### PR TITLE
refactor: sign up settings schema

### PR DESCRIPTION
**Pull request by xiaoyijun**
## Summary

---
> **NOTE:**
> This PR is large because schema modification affects the whole project from the back end to the front-end.
---

Change the `SignUp` schema from
```js
{
identifier: SignUpIdentifer;
password: boolean;
verify: boolean;
}
```
to
```js
{
identifiers: SignInIdentifier[];
password: boolean;
verify: boolean;
}
```

#### Schemas
- Remove the `SignUpIdentifier`.
- Change `identifier: SignUpIdentifier` to `identifiers: SignInIdentifier[]` in the `SignUp` type.
- Add an alteration script to alter the database.

#### Core
- Apply the new `SignUp` type to the legacy codes.
- Fix a test case called `should throw when there is no email connector and identifier is email or phone` in `packages/core/src/lib/sign-in-experience/sign-up.test.ts`, for we should not pass the `email` connector into the enabled connectors, and we should set the `verify` to `true` to avoid effect our connector test case.

### Main Flow
- Apply the new `SignUp` type to the legacy codes.
- Remove the `SignInExperienceSettings` type, for now, it's the same as `SignInExperienceResponse`.

### AC
- Apply the new `SignUp` type to the legacy codes.
- Refactor the `constants` file according to the new `SignUp` type.
- Split the `src/pages/SignInExperience/utilities` into `utils/form`, `utils/identifier`, and `utils/language`, for the code file become large after refactoring.

### Integration Test
- Apply the new `SignUp` type to the legacy codes.

## Testing

Test locally & tests passed.

### CODE REVIEW

**Review by wangsijie — APPROVED** [review]
Overall LGTM, better wait for another stamp.

And is there a better name to replace "SignInExperienceIdentifier"?

**Comment by xiaoyijun** [comment]
> Overall LGTM, better wait for another stamp.
> 
> And is there a better name to replace "SignInExperienceIdentifier"?

Any ideas? cc @logto-io/eng

File: packages/core/src/lib/sign-in-experience/sign-in.ts [hunk 1]
```diff
@@ -47,56 +47,31 @@ export const validateSignIn = (
...
-
-    case SignUpIdentifier.EmailOrSms: {
-      assertThat(
-        signIn.methods.some(({ identifier }) => identifier === SignInIdentifier.Email) &&
-          signIn.methods.some(({ identifier }) => identifier === SignInIdentifier.Sms),
-        new RequestError({
-          code: 'sign_in_experiences.miss_sign_up_identifier_in_sign_in',
-        })
-      );
+  if (signUp.identifiers.includes(SignInIdentifier.Username)) {
```
Comment (simeng-li) [inline packages/core/src/lib/sign-in-experience/sign-in.ts:50]: for (identifiers in signUp.identifiers) {}

Comment (xiaoyijun) [reply packages/core/src/lib/sign-in-experience/sign-in.ts]: Good call, updated.
Also, check the `identifers.includes` pattern in other codes, and refactor the codes in the `core/src/lib/sign-in-experience/sign-up.ts`.

File: packages/core/src/routes/interaction/utils/sign-in-experience-validation.ts [hunk 2]
```diff
@@ -59,7 +59,7 @@ export const identifierValidation = (
         if (
           'passcode' in identifier &&
           !verificationCode &&
-          ![SignUpIdentifier.Email, SignUpIdentifier.EmailOrSms].includes(signUp.identifier) &&
```
Comment (charIeszhao) [inline packages/core/src/routes/interaction/utils/sign-in-experience-validation.ts:62]: I don't think these expressions are equal. Should probably be:
```ts
!signUp.identifiers.includes(SignInIdentifier.Email) && !signUp.identifiers.includes(SignInIdentifier.EmailOrSms) &&
```

File: packages/core/src/routes/interaction/utils/sign-in-experience-validation.ts [hunk 3]
```diff
@@ -59,7 +59,7 @@ export const identifierValidation = (
...
+          !signUp.identifiers.includes(SignInIdentifier.Email) &&
```
Comment (charIeszhao) [inline packages/core/src/routes/interaction/utils/sign-in-experience-validation.ts:62]: Why check if signUp identifiers include a `sign in` identifier? I'm really confused

File: packages/core/src/routes/interaction/utils/sign-in-experience-validation.ts (hunk 2 above)
Comment (xiaoyijun) [reply packages/core/src/routes/interaction/utils/sign-in-experience-validation.ts:62]: This is new logic with a new type.
We don't have a `SignInIdentifer.EmailOrSms`, we use `includes(SignInIdentifier.Email)` to know the Sign-Up Email verification is enabled.

File: packages/console/src/pages/SignInExperience/components/SignUpAndSignInChangePreview/SignUpAndSignInDiffSection/index.tsx [hunk 4]
```diff
@@ -12,7 +14,11 @@ type Props = {
 
 const SignUpAndSignInDiffSection = ({ before, after, isAfter = false }: Props) => (
   <>
-    <SignUpDiffSection before={before.signUp} after={after.signUp} isAfter={isAfter} />
+    <SignUpDiffSection
+      before={signInExperienceParser.toLocalSignUp(before.signUp)}
+      after={signInExperienceParser.toLocalSignUp(after.signUp)}
```
Comment (charIeszhao) [inline packages/console/src/pages/SignInExperience/components/SignUpAndSignInChangePreview/SignUpAndSignInDiffSection/index.tsx:19]: These are a bit confusing. Can we pass `before` and `after` into this component and call the parser inside?

File: packages/console/src/pages/SignInExperience/components/SignUpAndSignInChangePreview/SignUpAndSignInDiffSection/utilities.ts [hunk 5]
```diff
@@ -1,12 +1,10 @@
...
+import type { SignUp } from '@logto/schemas';
 import { diff } from 'deep-object-diff';
 
 import type { SignInMethod, SignInMethodsObject } from '@/pages/SignInExperience/types';
 
-export const isSignUpDifferent = (
-  before: SignInExperience['signUp'],
-  after: SignInExperience['signUp']
-) => Object.keys(diff(before, after)).length > 0;
+export const isSignUpDifferent = (before: SignUp, after: SignUp) =>
```
Comment (charIeszhao) [inline packages/console/src/pages/SignInExperience/components/SignUpAndSignInChangePreview/SignUpAndSignInDiffSection/utilities.ts:6]: nit: `hasSignUpMethodChanged`?

File: packages/console/src/pages/SignInExperience/tabs/SignUpAndSignIn/SignUpForm.tsx [hunk 6]
```diff
@@ -56,7 +55,7 @@ const SignUpForm = () => {
       return;
     }
 
-    if (requiredVerifySignUpIdentifiers.includes(signUpIdentifier)) {
+    if (isVerificationRequiredSignUpIdentifiers(signUpIdentifier)) {
```
Comment (charIeszhao) [inline packages/console/src/pages/SignInExperience/tabs/SignUpAndSignIn/SignUpForm.tsx:58]: nit: maybe just `isSignUpIdentifierRequired`?

**Review by charIeszhao — APPROVED** [review]

File: packages/console/src/pages/SignInExperience/components/SignUpAndSignInChangePreview/SignUpAndSignInDiffSection/utilities.ts (hunk 5 above)
Comment (xiaoyijun) [reply packages/console/src/pages/SignInExperience/components/SignUpAndSignInChangePreview/SignUpAndSignInDiffSection/utilities.ts]: updated.

**Pull request merged**